### DISCUSSION TITLE
Reverse proxy NPM

### DISCUSSION THREAD

**Discussion by henkiewie** [Lissy93/dashy#427] [Help and Support] [answered]
Hi,
I started using dashy behind a reverse proxy with https. (Nginx Proxy manager) is there an option to make widgets like public ip-address or iframe  work?
regards Henk

**Comment by lissy93** [comment, marked as answer]
This should work in exactly the same way. What error are you getting?

If you're getting a CORS error, then you need to either set `useProxy: true` or set the correct allow headers - [here's more info](https://github.com/Lissy93/dashy/blob/master/docs/widgets.md#proxying-requests)

**Reply by henkiewie** [reply]
The error is "An error occurred, see the logs for more info. Unable to fetch IP info" Use proxy true doesn't help. same error. the embed widget works but iframe and public ipaddress do not. Because the nginx proxy manager config files are pretty dynamic (they are build through the webinterface) I think its better to leave them as is. There is an option to put extra values in the config, but that's not simple for me at least. I did manage to add the first variable proxy_set_header 'Access-Control-Allow-Origin' "https://dashy.mylocal.lan"; but the second Vary: order is still a bit of a mystery I can not seem to find the label to add this. It does not accept proxy_set_header

**Reply by lissy93** [reply]
I mean the error on the console, can you share the output please?

In terms of the iframe widget, access to content is something you need to set on the origin, it can not be controlled from within Dashy. Take a look at this guide in the [Troubleshooting Docs](https://github.com/Lissy93/dashy/blob/master/docs/troubleshooting.md#refused-to-connect-in-modal-or-workspace-view) for instructions. So you'll set these headers on the resource you're trying to access, not on Dashy.

**Reply by henkiewie** [reply]
I use the docker container that have no errors in the log. only then the compiling messages and watching for changes. I just keep using the http URL thta goes straight to the docker container for now. thanks for the help.

**Reply by lissy93** [reply]
Sorry, I meant the browser console. Whenever a widget has an error, the details will show up in the console. Without that info, or your config file, it's really hard to know what is happening. If you're unsure how to access it, [here is a short guide](https://github.com/Lissy93/dashy/blob/master/docs/troubleshooting.md#how-to-open-browser-console). Share this along with the relevant parts of your config file, and then I will know what's going on here.

If you're still not sure, then just [enable error reporting](https://github.com/Lissy93/dashy/blob/master/docs/contributing.md#enable-anonymous-bug-reports), and reproduce the error.

**Reply by henkiewie** [reply]
Ah OK. THe error in the browser is:
VM49:1 Mixed Content: The page at 'https://dashy.rpi-lite.lan/' was loaded over HTTPS, but requested an insecure XMLHttpRequest endpoint 'http://ip-api.com/json'. This request has been blocked; the content must be served over HTTPS. I fixed the iframe (same error) by changing the url to https.

**Reply by lissy93** [reply]
That is indeed just a common CORS error. Not an issue with Dashy, but a browser security feature which prevents the client-side app (Dashy) from from accessing resources on a remote origin, without that server's explicit permission (e.g. with headers like Access-Control-Allow-Origin). See the MDN Docs for more info: [Cross-Origin Resource Sharing](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS).

As mentioned above, the solution is just either setting correct headers or proxying through the server-side, all details are outlined in the docs.

If this still doesn't make sense, you can probably find more info online. Maybe this article will help:
https://medium.com/@dtkatz/3-ways-to-fix-the-cors-error-and-how-access-control-allow-origin-works-d97d55946d9

Hope that helps :)

**Reply by henkiewie** [reply]
OK thx for the help. Mixed content is a security issue for sure, but since its a local network it should be allowed. I found the right way to add security headers to my version of NGINX reverse proxy. The quest is well on the way. Thanks for putting me on the right track. Luctor et Emergo

**Reply by henkiewie** [reply]
Just as an afterburner; The ipaddress still doesn't work. Because finally you will get a 403 error because https is a paid service.
But the change to my website headers gave it grade A for security. :-).
This URL helped me make it; https://securityheaders.com/ And your glances widgets work great.
![Screenshot 2022-01-23 at 17 02 53](https://user-images.githubusercontent.com/74558874/150750135-a8829a6e-1229-4c11-9844-c54322e62033.png)

**Reply by lissy93** [reply]
Ohh you're right, I totally forgot about that. This is a bit of a flaw.
I really hate when companies put basic security features behind a paywall.

I'm going to find a new IP info API service, which doesn't have this requirement, Re: #437
Sorry about that, my mistake. I will fix this today or tomorrow, and update you when done :)

**Reply by henkiewie** [reply]
MAybe you can use glance? it has this info on top of the website.
(Debian GNU/Linux 10 32bit / Linux 5.10.92-v8+)
 - IP 172.18.0.2/16 Pub 217.xxx.xxx.xxx
Uptime: 3 days, 2:03:12

**Reply by lissy93** [reply]
Implemented in #438 :)

I've updated the IP address widget to use [ipapi.co](https://ipapi.co/) which should now work for you

I also created a Glances IP info widget too (`gl-ip-address`). But while testing I noticed that not all instances of Glances actually return any IP data.

**Reply by henkiewie** [reply]
Wow this ipapi.co know to muchI :-) But they are working fine after deploying a new Docker image. Thank you very much for the really Quick action.
